README.md
<a href="https://yeovilhospital.co.uk/">
	<img alttext="Yeovil District Hospital Logo" src="https://github.com/Fdawgs/ydh-logos/raw/HEAD/images/ydh-full-logo-transparent-background.svg" width="480" />
</a>

# Yeovil District Hospital NHS Foundation Trust - FHIR API

[![GitHub Release](https://img.shields.io/github/release/Fdawgs/ydh-fhir-api.svg)](https://github.com/Fdawgs/ydh-fhir-api/releases/latest/)
![Build Status](https://github.com/Fdawgs/ydh-fhir-api/workflows/CI/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/Fdawgs/ydh-fhir-api/badge.svg?branch=master)](https://coveralls.io/github/Fdawgs/ydh-fhir-api?branch=master)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat)](https://github.com/prettier/prettier)

> Yeovil District Hospital NHSFT's RESTful HL7® FHIR® API

## Intro

This is [Yeovil District Hospital NHSFT](https://yeovilhospital.co.uk/)'s RESTful HL7® FHIR® API, a Node.js application using the [Fastify](https://www.fastify.io/) web framework.

Background information detailing the history and purpose of this API can be found [here](./docs/history.md).

## Prerequisites

-   InterSystems IRIS ODBC35 ODBC Driver 2021
-   InterSystems IRIS for Health 2020
-   InterSystems TrakCare v2020
-   [Node.js](https://nodejs.org/en/) >=16.0.0 (if running outside of Docker)
-   [SQL Server](https://www.microsoft.com/en-gb/sql-server/sql-server-downloads) (As a Windows service/instance; SQL Server on Linux [does not support linked servers](https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-editions-and-components-2019?view=sql-server-ver15#Unsupported))

## Setup

Perform the following steps before deployment:

1. Clone or download the repo
2. Navigate to the project directory
3. Make a copy of `.env.template` in the root directory and rename it to `.env`
4. Configure the application using the environment variables in `.env`
5. Install InterSystems IRIS ODBC35 ODBC Driver
6. Follow the InterSystems documentation for [Defining an ODBC Data Source on Windows](https://docs.intersystems.com/irislatest/csp/docbook/DocBook.UI.Page.cls?KEY=BNETODBC_winodbc) as a System DSN

> **Note**
> You will need to create a database before using it in the `DB_CONNECTION_STRING` environment variable (this does not apply if using the included Docker Compose file to deploy)

> **Note**
> Set the following environment variables in `.env` to meet NHS Digital's recommendation to retain 6 months' worth of logs:
>
> -   `LOG_ROTATION_DATE_FORMAT="YYYY-MM-DD"`
> -   `LOG_ROTATION_FREQUENCY="daily"`
> -   `LOG_ROTATION_MAX_LOGS="180"`

## Deployment

### Standard Deployment

1. Run `npm ci --ignore-scripts --production` to install dependencies
2. Run `npm start`

The service should be up and running on the port set in the config. You should see output similar to the following in stdout or in the log file specified using the `LOG_ROTATION_FILENAME` environment variable:

```json
{
	"level": "info",
	"time": "2022-01-10T10:17:35.556Z",
	"pid": 18,
	"hostname": "MYCOMPUTER",
	"msg": "Connecting to MSSQL DB"
}
```

```json
{
	"level": "info",
	"time": "2022-01-10T10:17:35.558Z",
	"pid": 18,
	"hostname": "MYCOMPUTER",
	"msg": "MSSQL DB connection opened"
}
```

```json
{
	"level": "info",
	"time": "2022-01-10T10:17:35.760Z",
	"pid": 18,
	"hostname": "MYCOMPUTER",
	"msg": "Server listening at http://0.0.0.0:8204"
}
```

To quickly test it, use [Insomnia](https://insomnia.rest/) and import the example requests from `./test_resources/insomnia_test_requests.json`.

### Deploying Using Docker

This requires [Docker](https://www.docker.com) installed.

1. Run `docker compose up` (or `docker compose up -d` to run in background)

### Deploying Using PM2

If you are unable to deploy this into production using Docker, it is recommended that you use a process manager such as [PM2](https://pm2.keymetrics.io/).

1. Run `npm ci --ignore-scripts --production` to install dependencies
2. Run `npm i -g pm2` to install pm2 globally
3. Launch application with `pm2 start .pm2.config.js`
4. Check the application has been deployed using `pm2 list` or `pm2 monit`

#### To Install as a Windows Service:

If using a Microsoft Windows OS utilise [pm2-installer](https://github.com/jessety/pm2-installer) to install PM2 as a Windows service.

> **Note**
> PM2 will automatically restart the application if `.env` is modified.

## Known Issues and Caveats

Issues with InterSystems TrakCare PAS (used by YDH) and staff misuse of the PAS have affected how the data is presented in the endpoints and how searches can be performed.

### Data Quality

-   AllergyIntolerance resources:
    -   Unable to provide SNOMED codes for allergies and intolerances in AllergyIntolerance resources due to these being free text inputs in TrakCare
    -   Low recordings of allergy and intolerance data in TrakCare:
        -   350,513 non-deceased patients with records in TrakCare as of 2020-11-19
            -   34,405 patients have ‘No Known Allergy’ recorded (9.8%)
            -   13,139 patients have one or more allergies recorded (3.7%)
    -   Due to the above issues, Paul Foster (CCIO at YDH) on 2020-11-19 suggested we **do not provide AllergyIntolerance resources** (functionality is still present however)
-   Condition resources:
    -   **Unable to provide Condition resources** as conditions are held in SimpleCode, not TrakCare
-   DocumentReference resources:
    -   **Unable to provide DocumentReference resources** as these are held in Patient Centre, not TrakCare
-   Encounter resources:
    -   Discharge/end dates for outpatient Encounter resources are not provided due to poor data quality. Staff in outpatients misuse these input fields in TrakCare to mark when “all admin has been completed for that outpatient encounter” and not when the encounter actually finished
    -   Unable to provide clinician contact details for Encounter resources due to the following:
        -   In TrakCare a care provider has a mobile number field against them, but it is rarely populated
        -   There is not an internal contact number field in TrakCare
        -   If you want to reach say, a gynaecology consultant, you need to manually search a list on YDH’s intranet for their secretary’s extension number, and there is no indication as to how current the list is
        -   Teams do not have a contact number
-   Patient resources:
    -   Unable to provide SNOMED codes for religious affiliation as these are not in TrakCare (NHS Data Dictionary coding is provided, however)
    -   Unable to provide GP surgery/organisation `name` value for inline Organization FHIR resource as this data is also not held in TrakCare; using GP consultant name as a placeholder
    -   A sizeable number of patient records without postcodes

### Search Caveats

-   Every search request to a FHIR resource endpoint that is **NOT** the Patient FHIR resource endpoint **MUST** have a `patient` search parameter, this is to stop intentional or unintentional DOS attacks due to long-running SQL queries:
    -   `GET [baseUrl]/AllergyIntolerance?criticality=[code]` will return a 500 error
    -   `GET [baseUrl]/AllergyIntolerance?patient=[id]&criticality=[code]` will work

This is due to YDH not having direct control over the underlying databases of the PAS, so cannot add indexes or make appropriate performance tweaks to support searches without also filtering by patient.

## Contributing

Contributions are welcome, and any help is greatly appreciated!

See [the contributing guide](./CONTRIBUTING.md) for details on how to get started.
Please adhere to this project's [Code of Conduct](./CODE_OF_CONDUCT.md) when contributing.

## Acknowledgements

-   **Adam Wiles** (Clinical Coding Manager) - Procedure FHIR resource advice
-   **David Suckling** (Application Support Manager) - TrakCare database table structure and frontend support
-   [**Dunmail Hodkinson**](https://github.com/Dunmail) - HL7 FHIR STU3 specification adherence and best practices advice
-   **George Dampier** (Pharmacy Systems Manager) - MedicationStatement FHIR resource advice
-   **Jessica Male** (Application Support Testing Analyst) - TrakCare frontend support
-   [**Julian Matthews**](https://github.com/NHS-juju) - Bug reports
-   **John Simpson** (Advanced Clinical Pharmacist) - MedicationStatement FHIR resource advice
-   [**Mark Hunt**](https://github.com/nhsbandit) - JWT and JWKS integration and testing
-   **Michael McCormack** (Solutions Developer) - SQL query optimisation
-   [**Neil Hayes-Webster**](https://github.com/NeilHW-YDH) - SQL query optimisation
-   **Nicolas Noblet** (Senior Analyst) - SQL query optimisation, TrakCare database table structure advice
-   [**Will Jehring**](https://github.com/wjehring) - HL7 FHIR STU3 specification adherence advice, JWT testing

## License

`ydh-fhir-api` is licensed under the [MIT](./LICENSE) license.


CHANGELOG.md
# Changelog

## 1.0.0 (2022-07-14)


### ⚠ BREAKING CHANGES

* **plugins/jwt-jwks-auth:** obtain jwks_uri from openid endpoint
* scrypt now used for hashing stored bearer tokens

### Features

* add basic route and queries; add migrations ([#11](https://github.com/Fdawgs/ydh-fhir-api/issues/11)) ([24af6d6](https://github.com/Fdawgs/ydh-fhir-api/commit/24af6d6199ac9b45d905e4fea4251e5b2335f5b2))
* add basic stu3 flag route ([f69e887](https://github.com/Fdawgs/ydh-fhir-api/commit/f69e887d99e73dbba4abbcdcbf5fea59b67c4eac))


### Bug Fixes

* **plugins/jwt-jwks-auth:** respond with 401 error on missing header ([42541b0](https://github.com/Fdawgs/ydh-fhir-api/commit/42541b09337b5737fef693bc8e8017598d44a2ed))
* **plugins/serialize-json-to-xml:** 204 resps do not have content types ([7a1c825](https://github.com/Fdawgs/ydh-fhir-api/commit/7a1c82598164a377f33bc1ad8bd483727dc5277a))
* **routes:** time and timezone suffix are case-sensitive ([42a23e3](https://github.com/Fdawgs/ydh-fhir-api/commit/42a23e34d430d0b3f735793c5a7e68d5dcea0395))
* **routes:** timestamptz data type does not support `YYYY` patterns ([0234e04](https://github.com/Fdawgs/ydh-fhir-api/commit/0234e041aa1b197c6b4cadb5363779656dbab364))


### Continuous Integration

* add dependency-review job ([6f2f2c7](https://github.com/Fdawgs/ydh-fhir-api/commit/6f2f2c712dafddfc375cc975f35b4849ef3d114b))
* **automerge:** fix context ([fc73b47](https://github.com/Fdawgs/ydh-fhir-api/commit/fc73b474c5bba2a3d4f44ea3520c294ec1a8b0ef))
* **cd:** update org name for release-please-action ([70a8d9e](https://github.com/Fdawgs/ydh-fhir-api/commit/70a8d9eb439cf53d0fef329df663acd701a2363d))
* check `user.login` is dependabot instead of `actor` ([82c2c01](https://github.com/Fdawgs/ydh-fhir-api/commit/82c2c012133b0b8bc914650838ff8f8a7ed62087))
* **ci:** add node 18 to test matrix ([#51](https://github.com/Fdawgs/ydh-fhir-api/issues/51)) ([679b867](https://github.com/Fdawgs/ydh-fhir-api/commit/679b86731347400b75f651ec46ce4a90623bf651))
* **ci:** require `unit-tests` job to pass for `save-pr-number` job to run ([9881de0](https://github.com/Fdawgs/ydh-fhir-api/commit/9881de0fd576017f88f0fb6586d62c63a6021a9d))
* **ci:** use `lts/*` for node setup in lint job ([6049213](https://github.com/Fdawgs/ydh-fhir-api/commit/604921314e03e19f4e334165078cec37bf208474))
* **ci:** use `node-version` for node matrix key ([2bad596](https://github.com/Fdawgs/ydh-fhir-api/commit/2bad59627bf2ed21c59ae0d7f9c75385efe2ada0))
* **codeql:** only run on pr changes to `.html`, `.js`, and `.yml` files ([2c19d3e](https://github.com/Fdawgs/ydh-fhir-api/commit/2c19d3e843f01159717726ce8c96e669a6249bf4))
* **codeql:** resolve missing analyses ([96fa85a](https://github.com/Fdawgs/ydh-fhir-api/commit/96fa85aaf03e6a78c4d8629620e1e1b9bc653e08))
* **codeql:** specify which files to scan during analysis ([779a4e6](https://github.com/Fdawgs/ydh-fhir-api/commit/779a4e6ebb623ad7cb0c5ce0ed3150ba14d28c3b))
* **link-check:** replace `npx linkinator` call with github action ([de77969](https://github.com/Fdawgs/ydh-fhir-api/commit/de779696314709b946c7221818082293a297150a))
* only trigger dependency-review on pr ([4268fc8](https://github.com/Fdawgs/ydh-fhir-api/commit/4268fc8ca5e615297960c74937f81e63968fc588))
* re-add docker tests ([67a5666](https://github.com/Fdawgs/ydh-fhir-api/commit/67a56660ed9094520e0958883813535f394a4e36))
* reduce workflow permissions to minimum ([8617110](https://github.com/Fdawgs/ydh-fhir-api/commit/86171100582656757a7acfb1fd983b4188c958c3))
* remove git credentials after checkout ([#54](https://github.com/Fdawgs/ydh-fhir-api/issues/54)) ([97b6700](https://github.com/Fdawgs/ydh-fhir-api/commit/97b67000ac1962e63ec6a61b5e7836661dec395e))
* replace workflow-run-cleanup-action with github concurrency ([f75df22](https://github.com/Fdawgs/ydh-fhir-api/commit/f75df225d0a9076f1509d60c2711cacf28e911d7))
* use shorter arg aliases for lockfile lint step ([de31121](https://github.com/Fdawgs/ydh-fhir-api/commit/de31121ef5cb034d003370ce7f67e3a8dabbdbb6))
* validate that resolved url matches the package name ([a68bb72](https://github.com/Fdawgs/ydh-fhir-api/commit/a68bb7256ebbbc1b351b493ed5f4410c6cb5a466))


### Improvements

* access `fs/promises` api via newer route ([d330ac2](https://github.com/Fdawgs/ydh-fhir-api/commit/d330ac2ebfa89e294fcf5b4e816af033da059adf))
* **config:** replace `forEach()` with `map()` ([#55](https://github.com/Fdawgs/ydh-fhir-api/issues/55)) ([25f337b](https://github.com/Fdawgs/ydh-fhir-api/commit/25f337b86c3dd523fdd7c71d13b9646344240354))
* **plugins/hashed-bearer-auth:** replace blocking `for()` loop ([59feee4](https://github.com/Fdawgs/ydh-fhir-api/commit/59feee47f1dab8d79ca1b435458a2b70cc246733))
* **plugins/jwt-jwks-auth:** obtain jwks_uri from openid endpoint ([ab64260](https://github.com/Fdawgs/ydh-fhir-api/commit/ab64260393a00ba51ad6270113f802d897d72348))
* **plugins/jwt-jwks-auth:** remove `await` in `Promise.any()` ([75e084c](https://github.com/Fdawgs/ydh-fhir-api/commit/75e084c4a170a793b0028b7480571a2861e856c4))
* replace pbkdf2 with scrypt ([#57](https://github.com/Fdawgs/ydh-fhir-api/issues/57)) ([0c440e4](https://github.com/Fdawgs/ydh-fhir-api/commit/0c440e40fbc8f8bf1f5a075663c1c80125637f33))


### Miscellaneous

* **.eslintrc:** enable `plugin:jest/style` rules ([0a3cdda](https://github.com/Fdawgs/ydh-fhir-api/commit/0a3cddaf89898901cba9809065325b2622977964))
* **.github/codeql-config:** remove quotation marks ([0c03109](https://github.com/Fdawgs/ydh-fhir-api/commit/0c03109fa61f10784245e21cbecb743d349b95ab))
* **.github/funding:** remove comments ([3fc1cc2](https://github.com/Fdawgs/ydh-fhir-api/commit/3fc1cc27c23754e2effc2503a4d283f354c4b13f))
* **.github/workflows/link-check:** use `skip` input ([31f6f75](https://github.com/Fdawgs/ydh-fhir-api/commit/31f6f75549c2a4f582310ed2449a5eafe65cc9a1))
* **.github:** add `FUNDING.yml` ([cafb845](https://github.com/Fdawgs/ydh-fhir-api/commit/cafb84585d0ffa08f124efa57f38070d976d232b))

src/routes/STU3/Flag/schema.js
const S = require("fluent-json-schema");

const tags = ["STU3"];

const dateTimeSearchPattern =
	/^(?:eq|ne|ge|le|gt|lt|sa|eb|ap|)\d{4}(?:-[0-1]\d-[0-3]\d(?:t(?:[0-2]\d:[0-5]\d:[0-5]\d|23:59:60)(?:\.\d+)?(?:z|[+-]\d\d(?::?\d\d)?))?)?$/im;

const dateTimeSearchPatternExamples = [
	"2022",
	"2022-01-13",
	"2022-01-13T00:00:01Z",
	"2022-01-13T00:00:01.001Z",
	"2022-01-13T00:00:01+01:00",
	"ge2022",
	"ge2022-01-13",
	"ge2022-01-13T00:00:01Z",
	"ge2022-01-13T00:00:01.001Z",
	"ge2022-01-13T00:00:01+01:00",
];

/**
 * Fastify uses AJV for JSON Schema Validation,
 * see https://www.fastify.io/docs/latest/Validation-and-Serialization/
 *
 * Input validation protects against XSS, HPP, and most injection attacks.
 */
const stu3FlagReadSchema = {
	tags,
	summary: "Read an HL7 FHIR STU3 Flag resource",
	description: "Return a single HL7 FHIR STU3 Flag resource.",
	operationId: "getReadStu3Flag",
	produces: ["application/fhir+json", "application/fhir+xml"],
	params: S.object()
		.prop(
			"id",
			S.string()
				.description("Logical id of artifact")
				.examples(["999999-99"])
				.pattern(/^[\d-]+$/m)
		)
		.required(["id"]),
	response: {
		// TODO: build out 200 response object
		// 200: S.object(),
		401: S.ref("responses#/definitions/unauthorized").description(
			"Unauthorized"
		),
		// 404: S.ref(
		// 	"responses#/definitions/notFoundOperationOutcome"
		// ).description("Not Found"),
		406: S.ref("responses#/definitions/notAcceptable").description(
			"Not Acceptable"
		),
		429: S.ref("responses#/definitions/tooManyRequests").description(
			"Too Many Requests"
		),
		500: S.ref("responses#/definitions/internalServerError").description(
			"Internal Server Error"
		),
		503: S.ref("responses#/definitions/serviceUnavailable").description(
			"Service Unavailable"
		),
	},
};

const stu3FlagSearchSchema = {
	tags,
	summary: "Search HL7 FHIR STU3 Flag resources",
	description: "Return a bundle of HL7 FHIR STU3 Flag resources.",
	operationId: "getSearchStu3Flag",
	produces: ["application/fhir+json", "application/fhir+xml"],
	query: S.object()
		.prop(
			"date",
			S.anyOf([
				S.string()
					.description("Time period when flag is active")
					.examples(dateTimeSearchPatternExamples)
					.pattern(dateTimeSearchPattern),
				S.array()
					.items(
						S.string()
							.description("Time period when flag is active")
							.examples(dateTimeSearchPatternExamples)
							.pattern(dateTimeSearchPattern)
					)
					.minItems(2)
					.maxItems(2)
					.uniqueItems(true),
			])
		)
		.prop(
			"patient",
			S.number().description(
				"The identity of a subject to list flags for"
			)
		)
		.prop(
			"patient.identifier",
			S.string().description(
				"The identity of a subject to list flags for"
			)
		)
		.prop("status", S.string().enum(["active", "inactive"])),
	response: {
		// TODO: build out 200 response object
		// 200: S.object(),
		401: S.ref("responses#/definitions/unauthorized").description(
			"Unauthorized"
		),
		// 404: S.ref(
		// 	"responses#/definitions/notFoundOperationOutcome"
		// ).description("Not Found"),
		406: S.ref("responses#/definitions/notAcceptable").description(
			"Not Acceptable"
		),
		429: S.ref("responses#/definitions/tooManyRequests").description(
			"Too Many Requests"
		),
		500: S.ref("responses#/definitions/internalServerError").description(
			"Internal Server Error"
		),
		503: S.ref("responses#/definitions/serviceUnavailable").description(
			"Service Unavailable"
		),
	},
};

module.exports = {
	stu3FlagReadSchema,
	stu3FlagSearchSchema,
};
